horovod/torch/mpi_ops_v2.cc

```diff
@@ -372,13 +372,17 @@ int DoBroadcastCudaOnCPU(::torch::Tensor tensor, ::torch::Tensor output, int roo
   return handle;
 }
 
-int DoAlltoall(::torch::Tensor tensor, ::torch::Tensor splits, ::torch::Tensor output, const std::string& name) {
+int DoAlltoall(::torch::Tensor tensor, ::torch::Tensor splits,
+               ::torch::Tensor output, ::torch::Tensor output_received_splits,
+               const std::string& name) {
   ThrowIfError(common::CheckInitialized());
 
   auto device = GetDeviceID(tensor);
   auto ready_event = RecordReadyEvent(device);
   auto hvd_tensor = std::make_shared<TorchTensor>(tensor);
   auto hvd_context = std::make_shared<TorchOpContext>(device, output);
+  assert(GetDeviceID(output_received_splits) == CPU_DEVICE_ID);
```

Comment:
@romerojosh, here I wasn't sure how to best deal with the possibility that output_received_splits might be a GPU tensor (which could be caused by a bug on the Python side).

Reply:
I think in the case where `output_received_splits` is a GPU tensor, we should make a temporary CPU output tensor, use that in the backend, and then copy the CPU output tensor to the GPU output tensor during the callback, similar to what is done inthe `*CudaOnCPU`routines. What do you think?

Reply:
Thanks, that sounds reasonable. I'll look into.

Reply:
OK, I have changed the PyTorch API such that if the incoming `splits` happen to be on GPU, the `received_splits` will be put on GPU, too. That's probably the least surprising behavior.

Reply:
I agree. Thanks for updating the PR!